Support IAM Roles for Service Accounts

Amazon recently released support for assigning IAM roles to Service Accounts (https://aws.amazon.com/blogs/opensource/introducing-fine-grained-iam-roles-service-accounts/). 

In order to take advantage of this, the AWS SDK needs to be upgraded (The Go SDK needs to be a minimum of version 1.23.13), and we need a way to indicate the IAM role to be used, so it can be added as an annotation to the service account(s) used by the operator.

Assigning IAM roles through this method is easier and more secure than keeping long-running IAM users around, which have fixed credentials and [ideally] need them rotated periodically.

I figured out what's going on here. 

First, I rebuilt the ansible-operator image using a modified hive-configmap.yaml that enabled the debug logging for s3a. This confirmed that s3a was using only the BasicAWSCredentialsProvider, EnvironmentVariableCredentialsProvider, and InstanceProfileCredentialsProvider - not the new WebIdentityTokenCredentialsProvider. 

Then I edited _hadoop_config_helpers.tpl to force the use of the new provider, by setting 
```
  <property>
      <name>fs.s3a.aws.credentials.provider</name>
      <value>com.amazonaws.auth.WebIdentityTokenCredentialsProvider</value>
  </property>
```
which then caused it to throw the following stack trace:
```
Caused by: java.io.IOException: com.amazonaws.auth.WebIdentityTokenCredentialsProvider constructor exception.  A class specified in fs.s3a.aws.credentials.provider must provide a public constructor accepting Configuration, or a public factory method named getInstance that accepts no arguments, or a public default constructor.
        at org.apache.hadoop.fs.s3a.S3AUtils.createAWSCredentialProvider(S3AUtils.java:661) ~[hadoop-aws-3.1.1.jar:?]
        at org.apache.hadoop.fs.s3a.S3AUtils.createAWSCredentialProviderSet(S3AUtils.java:566) ~[hadoop-aws-3.1.1.jar:?]
```

And sure enough, WebIdentityTokenCredentialsProvider offers neither of those things. Looking at how it's used in the AWS SDK (https://github.com/aws/aws-sdk-java/blob/53683b836dec55226f45060dd5af3f00259650c3/aws-java-sdk-core/src/main/java/com/amazonaws/auth/DefaultAWSCredentialsProviderChain.java) you can see that it uses `.create()` rather than `.getInstance()`

So I guess that puts me back to editing the Hadoop fork to use the new provider from the new SDK.

--- a/charts/openshift-metering/templates/hadoop/hdfs-serviceaccount.yaml
+++ b/charts/openshift-metering/templates/hadoop/hdfs-serviceaccount.yaml
@@ -3,4 +3,8 @@ kind: ServiceAccount
 metadata:
   name: hdfs
   labels:
-      app: hdfs
+    app: hdfs
+{{- if .Values.hadoop.spec.hdfs.namenode.annotations }}
+  annotations:
+{{ toYaml .Values.hadoop.spec.hdfs.namenode.annotations | indent 4 }}
+{{- end }}

--- a/charts/openshift-metering/templates/hive/hive-serviceaccount.yaml
+++ b/charts/openshift-metering/templates/hive/hive-serviceaccount.yaml
@@ -2,3 +2,9 @@ apiVersion: v1
 kind: ServiceAccount
 metadata:
   name: hive
+  labels:
+    app: hive
+{{- if .Values.hive.spec.annotations }}
+  annotations:
+{{ toYaml .Values.hive.spec.annotations | indent 4 }}
+{{- end }}

--- a/charts/openshift-metering/templates/presto/presto-serviceaccount.yaml
+++ b/charts/openshift-metering/templates/presto/presto-serviceaccount.yaml
@@ -2,3 +2,9 @@ apiVersion: v1
 kind: ServiceAccount
 metadata:
   name: presto
+  labels:
+    app: presto
+{{- if .Values.presto.spec.annotations }}
+  annotations:
+{{ toYaml .Values.presto.spec.annotations | indent 4 }}
+{{- end }}

--- a/charts/openshift-metering/templates/reporting-operator/reporting-operator-serviceaccount.yaml
+++ b/charts/openshift-metering/templates/reporting-operator/reporting-operator-serviceaccount.yaml
@@ -5,7 +5,10 @@ metadata:
   name: reporting-operator
   labels:
     app: reporting-operator
-{{- if and $operatorValues.spec.authProxy.enabled $operatorValues.spec.route.enabled }}
   annotations:
+{{- if and $operatorValues.spec.authProxy.enabled $operatorValues.spec.route.enabled }}
     serviceaccounts.openshift.io/oauth-redirectreference.reporting-operator: '{"kind":"OAuthRedirectReference","apiVersion":"v1","reference":{"kind":"Route","name":"{{ $operatorValues.spec.route.name }}"}}'
+{{- end }}
+{{- if $operatorValues.spec.annotations }}
+{{ toYaml $operatorValues.spec.annotations | indent 4 }}
 {{- end }}

